compiler/rustc_mir_dataflow/src/impls/single_enum_variant.rs

```diff
@@ -0,0 +1,171 @@
+            } else {
+                state.remove(lhs_local);
+            }
+        };
+    }
+}
+
+impl<'tcx> Analysis<'tcx> for SingleEnumVariant<'_, 'tcx> {
+    fn apply_statement_effect(
+        &self,
+        state: &mut Self::Domain,
+        statement: &Statement<'tcx>,
+        loc: Location,
+    ) {
+        let (place, fact) = match &statement.kind {
+            StatementKind::Deinit(box place) => (place, None),
+            StatementKind::SetDiscriminant { box place, variant_index } => {
+                (place, Some(*variant_index))
+            }
+            StatementKind::Assign(box (lhs, Rvalue::Use(op))) => {
+                return self.assign(state, lhs, op, loc);
+            }
+            /* may alias/mutate RHS need to specify that it is no longer a single value */
+            StatementKind::Assign(box (
+                _,
+                Rvalue::Ref(_, BorrowKind::Mut { .. }, rhs)
+                | Rvalue::AddressOf(Mutability::Mut, rhs),
+            )) => (rhs, None),
+            StatementKind::CopyNonOverlapping(box ref copy) => {
+                let place = match &copy.dst {
+                    Operand::Copy(p) | Operand::Move(p) => p,
+                    _ => return,
+                };
+                (place, None)
+            }
+            _ => return,
+        };
+        if !self.is_tracked(place) {
+            return;
+        }
+        let Some(local) = place.local_or_deref_local() else { return };
+        if let Some(fact) = fact {
+            state.insert(local, loc, fact);
+        } else {
+            state.remove(local);
+        }
+    }
+    fn apply_terminator_effect(
+        &self,
+        state: &mut Self::Domain,
+        terminator: &Terminator<'tcx>,
+        loc: Location,
+    ) {
+        match &terminator.kind {
+            TerminatorKind::DropAndReplace { place, value, .. } => {
+                self.assign(state, place, value, loc)
+            }
+            TerminatorKind::Drop { place, .. } if self.is_tracked(place) => {
+                let Some(local) = place.local_or_deref_local() else { return };
+                state.remove(local);
+            }
+            _ => {}
+        }
+    }
+
+    fn apply_call_return_effect(
+        &self,
+        _: &mut Self::Domain,
+        _: BasicBlock,
+        _: CallReturnPlaces<'_, 'tcx>,
+    ) {
+    }
+
+    fn apply_switch_int_edge_effects(
+        &self,
+        _block: BasicBlock,
+        discr: &Operand<'tcx>,
+        apply_edge_effects: &mut impl SwitchIntEdgeEffects<Self::Domain>,
+    ) {
+        let Some(place) = discr.place() else { return };
+        if !self.is_tracked(&place) {
+            return;
+        }
+        let Some(local) = place.local_or_deref_local() else { return };
+        apply_edge_effects.apply(|state, target| {
+            // This probably isn't right, need to check that it fits.
+            let new_fact = target.value.map(|v| VariantIdx::from_u32(v as u32));
```

Comment:
Yeah, this is unfortunately an easy footgun. The variant index and discriminant value are not the same in general, see <https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/struct.Ty.html#method.discriminant_for_variant> and related methods

Reply:
Is this a discriminant index or is it a variant index? I can't tell (and don't remember what I thought)

Reply:
(nit: discriminant's aren't indexes, just values). Your opt is (correctly) storing variant indexes. The values in `target` are just compared to the value in the operand as integers. I guess this bug is probably just the same bug as the one below - it probably makes more sense to fix that first and then revisit this